packages/koa-react-sidecar/src/render/render.tsx

```diff
@@ -1,46 +1,106 @@
 import {renderToString} from 'react-dom/server';
-import {Html, HtmlManager, HtmlContext} from '@shopify/react-html/server';
+import {
+  Html,
+  HtmlManager,
+  HtmlContext,
+  Serialize,
+} from '@shopify/react-html/server';
 import {
   applyToContext,
   NetworkContext,
   NetworkManager,
 } from '@shopify/react-network/server';
+import {createApolloBridge} from '@shopify/react-effect-apollo';
 import {extract} from '@shopify/react-effect/server';
+import {AsyncAssetContext, AsyncAssetManager} from '@shopify/react-async';
+import {Omit} from '@shopify/useful-types';
 import {Header} from '@shopify/react-network';
+import {getAssets} from '@shopify/sewing-kit-koa';
+import {createGraphQLClient, GraphQLClientOptions} from '../graphql-client';
+import {getLogger} from '../logger';
+
+export type RenderContext = Context & {
+  graphQLClientOptions?: GraphQLClientOptions;
+  graphQLClient?: ApolloClient<{}>;
+};
 
-export type RenderContext = Context & {};
+export interface CreateRenderOptions {
+  render: (ctx: RenderContext) => React.ReactElement<any>;
+  graphQLClientOptions?: Omit<GraphQLClientOptions, 'host'>;
+}
 
-const logger = console;
+export function createRender(options: CreateRenderOptions) {
+  const {render, graphQLClientOptions} = options;
 
-export function createRender(
-  renderFunction: (ctx: RenderContext) => React.ReactElement<any>,
-) {
-  return async function render(ctx: RenderContext) {
-    const app = renderFunction(ctx);
+  return async function renderFunction(ctx: RenderContext) {
+    const logger = getLogger(ctx);
+    const assets = getAssets(ctx);
     const networkManager = new NetworkManager();
     const htmlManager = new HtmlManager();
+    const graphQLClient =
+      graphQLClientOptions &&
+      createGraphQLClient({host: ctx.host, ...graphQLClientOptions});
+    const ApolloBridge = createApolloBridge();
+    const asyncAssetManager = new AsyncAssetManager();
+    const app = render({...ctx, graphQLClientOptions, graphQLClient});
 
     try {
       await extract(app, {
         decorate(app) {
           return (
-            <NetworkContext.Provider value={networkManager}>
-              <HtmlContext.Provider value={htmlManager}>
-                {app}
-              </HtmlContext.Provider>
-            </NetworkContext.Provider>
+            <HtmlContext.Provider value={htmlManager}>
+              <AsyncAssetContext.Provider value={asyncAssetManager}>
+                <NetworkContext.Provider value={networkManager}>
+                  <ApolloBridge>{app}</ApolloBridge>
+                </NetworkContext.Provider>
+              </AsyncAssetContext.Provider>
+            </HtmlContext.Provider>
           );
         },
+        afterEachPass({renderDuration, resolveDuration, index, finished}) {
+          const pass = `Pass number ${index} ${
+            finished ? ' (this was the final pass)' : ''
+          }`;
+          const rendering = `Rendering ${renderDuration}ms`;
```

Comment:
These are kind of confusing / misleading since we just told them a pass finished but `Rendering` makes it sound like we are talking about one that is ongoing.  If I was reading I might think that it was talking about the next one.